test/pattern_matcher_test.py

```diff
@@ -0,0 +1,259 @@
+        assert PatternMatcher(patterns).matches(text) is expected
+
+
+def test_clean_patterns():
+    patterns = ["docs", "config"]
+    pm = PatternMatcher(patterns)
+    cleaned = pm.patterns
+    assert len(cleaned) == 2
+
+
+def test_clean_patterns_strip_empty_patterns():
+    patterns = ["docs", "config", ""]
+    pm = PatternMatcher(patterns)
+    cleaned = pm.patterns
+    assert len(cleaned) == 2
+
+
+def test_clean_patterns_exception_flag():
+    patterns = ["docs", "!docs/README.md"]
+    pm = PatternMatcher(patterns)
+    assert pm.exclusions
+
+
+def test_clean_patterns_leading_space_trimmed():
+    patterns = ["docs", "  !docs/README.md"]
+    pm = PatternMatcher(patterns)
+    assert pm.exclusions
+
+
+def test_clean_patterns_trailing_space_trimmed():
+    patterns = ["docs", "!docs/README.md  "]
+    pm = PatternMatcher(patterns)
+    assert pm.exclusions
+
+
+def test_clean_patterns_error_single_exception():
+    patterns = ["!"]
+    try:
+        PatternMatcher(patterns)
+    except ValueError as e:
+        assert str(e) == 'Illegal exclusion pattern: "!"'
+
+
+def test_match():
+    match_tests = [
+        ("abc", "abc", True, None),
+        ("*", "abc", True, None),
+        ("*c", "abc", True, None),
+        ("a*", "a", True, None),
+        ("a*", "abc", True, None),
+        ("a*", "ab/c", True, None),
+        ("a*/b", "abc/b", True, None),
+        ("a*/b", "a/c/b", False, None),
+        ("a*b*c*d*e*/f", "axbxcxdxe/f", True, None),
+        ("a*b*c*d*e*/f", "axbxcxdxexxx/f", True, None),
+        ("a*b*c*d*e*/f", "axbxcxdxe/xxx/f", False, None),
+        ("a*b*c*d*e*/f", "axbxcxdxexxx/fff", False, None),
+        ("a*b?c*x", "abxbbxdbxebxczzx", True, None),
+        ("a*b?c*x", "abxbbxdbxebxczzy", False, None),
+        ("ab[c]", "abc", True, None),
+        ("ab[b-d]", "abc", True, None),
+        ("ab[e-g]", "abc", False, None),
+        ("ab[^c]", "abc", False, None),
+        ("ab[^b-d]", "abc", False, None),
+        ("ab[^e-g]", "abc", True, None),
+        ("a\\*b", "a*b", True, None),
+        ("a\\*b", "ab", False, None),
+        ("a?b", "a☺b", True, None),
+        ("a[^a]b", "a☺b", True, None),
+        ("a???b", "a☺b", False, None),
+        ("a[^a][^a][^a]b", "a☺b", False, None),
+        ("[a-ζ]*", "α", True, None),
+        ("*[a-ζ]", "A", False, None),
+        ("a?b", "a/b", False, None),
+        ("a*b", "a/b", False, None),
+        ("[\\]a]", "]", True, None),
+        ("[\\-]", "-", True, None),
+        ("[x\\-]", "x", True, None),
+        ("[x\\-]", "-", True, None),
+        ("[x\\-]", "z", False, None),
+        ("[\\-x]", "x", True, None),
+        ("[\\-x]", "-", True, None),
+        ("[\\-x]", "a", False, None),
+        # These do not return errors because the Python re.compile() method does
+        # not raise an error on invalid syntax like Go does. We can omit the
+        # tests though since it doesn't affect behavior on _correct_ syntax.
```

Comment:
FYI

Reply:
Is there a possibility that a user could be expecting this to not match something that ends up getting matched? What's the behavior on invalid syntax for `re`? I'd hope it would match nothing.

Reply:
Go ends up throwing an error, the invalid syntax is for example a mismatched bracket which just gets interpreted literally.

Reply:
Nice so they can't bring an invalid dockerfile anyway